File: solutions/scala/005-longest-palindromic-substring.scala
/**
 * Problem 5: Longest Palindromic Substring (LeetCode 5)
 * Difficulty: Med
 * Language: Scala
 */
def longestPalindrome(s: String): String = {
  def expand(l: Int, r: Int): String = {
    var (ll, rr) = (l, r)
    while (ll >= 0 && rr < s.length && s(ll) == s(rr)) { ll -= 1; rr += 1 }
    s.substring(ll + 1, rr)
  }
  s.indices.foldLeft("") { (best, i) =>
    val odd  = expand(i, i)
    val even = expand(i, i + 1)
    val candidate = if (odd.length >= even.length) odd else even
    if (candidate.length > best.length) candidate else best
  }
}

File: solutions/elixir/005-longest-palindromic-substring.ex
# Problem 5: Longest Palindromic Substring (LeetCode 5)
# Difficulty: Med
# Language: Elixir
# 
defmodule LongestPalindrome do
  def longest_palindrome(s) do
    chars = String.graphemes(s)
    len = length(chars)

    Enum.reduce(0..(len - 1), "", fn i, best ->
      odd  = expand(chars, i, i)
      even = expand(chars, i, i + 1)
      candidate = if String.length(odd) >= String.length(even), do: odd, else: even
      if String.length(candidate) > String.length(best), do: candidate, else: best
    end)
  end

  defp expand(chars, l, r) when l < 0 or r >= length(chars), do: ""
  defp expand(chars, l, r) do
    if Enum.at(chars, l) == Enum.at(chars, r) do
      expand(chars, l - 1, r + 1)
    else
      Enum.slice(chars, (l + 1)..(r - 1)) |> Enum.join()
    end
  end
end
